Development/nmos/configuration_api.cpp

```diff
@@ -0,0 +1,654 @@
+
+    web::http::experimental::listener::api_router make_configuration_api(node_model& model, web::http::experimental::listener::route_handler validate_authorization, get_control_protocol_class_descriptor_handler get_control_protocol_class_descriptor, get_control_protocol_datatype_descriptor_handler get_control_protocol_datatype_descriptor, get_control_protocol_method_descriptor_handler get_control_protocol_method_descriptor, control_protocol_property_changed_handler property_changed, slog::base_gate& gate)
+    {
+        using namespace web::http::experimental::listener::api_router_using_declarations;
+
+        api_router configuration_api;
+
+        configuration_api.support(U("/?"), methods::GET, [](http_request req, http_response res, const string_t&, const route_parameters&)
+        {
+            set_reply(res, status_codes::OK, nmos::make_sub_routes_body({ U("x-nmos/") }, req, res));
+            return pplx::task_from_result(true);
+        });
+
+        configuration_api.support(U("/x-nmos/?"), methods::GET, [](http_request req, http_response res, const string_t&, const route_parameters&)
+        {
+            set_reply(res, status_codes::OK, nmos::make_sub_routes_body({ U("configuration/") }, req, res));
+            return pplx::task_from_result(true);
+        });
+
+        if (validate_authorization)
+        {
+            configuration_api.support(U("/x-nmos/") + nmos::patterns::configuration_api.pattern + U("/?"), validate_authorization);
+            configuration_api.support(U("/x-nmos/") + nmos::patterns::configuration_api.pattern + U("/.*"), validate_authorization);
+        }
+
+        const auto versions = with_read_lock(model.mutex, [&model] { return nmos::is14_versions::from_settings(model.settings); });
+        configuration_api.support(U("/x-nmos/") + nmos::patterns::configuration_api.pattern + U("/?"), methods::GET, [versions](http_request req, http_response res, const string_t&, const route_parameters&)
+        {
+            set_reply(res, status_codes::OK, nmos::make_sub_routes_body(nmos::make_api_version_sub_routes(versions), req, res));
+            return pplx::task_from_result(true);
+        });
+
+        configuration_api.mount(U("/x-nmos/") + nmos::patterns::configuration_api.pattern + U("/") + nmos::patterns::version.pattern, make_unmounted_configuration_api(model, get_control_protocol_class_descriptor, get_control_protocol_datatype_descriptor, get_control_protocol_method_descriptor, property_changed, gate));
+
+        return configuration_api;
+    }
+
+    namespace details
+    {
+        namespace fields
+        {
+            const web::json::field_as_string_or level{ U("level"), {} };
+            const web::json::field_as_string_or index{ U("index"), {} };
+            const web::json::field_as_string_or describe{ U("describe"), {} };
+        }
```

Comment:
Can be removed, i think. The code below uses `nmos::fields::nc::...`

Reply:
Thanks, @maweit; well spotted; update the code next.